### ISSUE TITLE
Ganeti virtio-net multiqueue support is (was?) broken

### ISSUE THREAD

**Issue by yiannist** [ganeti/ganeti#1268]
There has been a discussion recently about the multiqueue support in Ganeti's [mailing list](https://groups.google.com/forum/#!topic/ganeti/8_LI7SQqPdg).

**TL;DR:** This feature is broken because, throughout Ganeti's code-base, it is assumed that there is an one-to-one relationship between taps and VM NICs.

Multiqueue support was introduced in 47eca2c618107b09c8474539963fb05f0c692003. Since then, QEMU had some changes and thus the following patch is needed (thanks Andrew!):
```diff
diff --git a/lib/hypervisor/hv_kvm/__init__.py b/lib/hypervisor/hv_kvm/__init__.py
index 9fdae5383..62d7ab9e5 100644
--- a/lib/hypervisor/hv_kvm/__init__.py
+++ b/lib/hypervisor/hv_kvm/__init__.py
@@ -569,7 +569,7 @@ class KVMHypervisor(hv_base.BaseHypervisor):
   _QMP_RE = re.compile(r"^-qmp\s", re.M)
   _SPICE_RE = re.compile(r"^-spice\s", re.M)
   _VHOST_RE = re.compile(r"^-net\s.*,vhost=on|off", re.M)
-  _VIRTIO_NET_QUEUES_RE = re.compile(r"^-net\s.*,fds=x:y:...:z", re.M)
+  _VIRTIO_NET_QUEUES_RE = re.compile(r"^-net(dev)?\s.*,fds=x:y:...:z", re.M)
   _ENABLE_KVM_RE = re.compile(r"^-enable-kvm\s", re.M)
   _DISABLE_KVM_RE = re.compile(r"^-disable-kvm\s", re.M)
   _NETDEV_RE = re.compile(r"^-netdev\s", re.M)
@@ -1835,9 +1835,9 @@ class KVMHypervisor(hv_base.BaseHypervisor):
           # Check for multiqueue virtio-net support.
           if self._VIRTIO_NET_QUEUES_RE.search(kvmhelp):
             # As advised at http://www.linux-kvm.org/page/Multiqueue formula
-            # for calculating vector size is: vectors=2*N+1 where N is the
+            # for calculating vector size is: vectors=2*N+2 where N is the
             # number of queues (HV_VIRTIO_NET_QUEUES).
-            nic_extra_str = ",mq=on,vectors=%d" % (2 * virtio_net_queues + 1)
+            nic_extra_str = ",mq=on,vectors=%d" % (2 * virtio_net_queues + 2)
             update_features["mq"] = (True, virtio_net_queues)
           else:
             raise errors.HypervisorError("virtio_net_queues is configured"
```
(Note that the [QEMU documentation](https://www.linux-kvm.org/page/Multiqueue#Enable_MQ_feature) states that the vector formula should be `2*N+2`: N for tx queues, N for rx queues, 1 for config, and one for possible control vq.)

With this patch, and a 4.9 kernel in the guest VM (although 3.x should work too, 3.16 didn't work for us), we should be able to see the queues:
```
root@stretch-test:~# ethtool -l eth0
Channel parameters for eth0:
Pre-set maximums:
RX:		0
TX:		0
Other:		0
Combined:	4
Current hardware settings:
RX:		0
TX:		0
Other:		0
Combined:	1

root@stretch-test:~# ls /sys/class/net/eth0/queues/
rx-0  tx-0

root@stretch-test:~# ethtool -L eth0 combined 4

root@stretch-test:~# ethtool -l eth0
Channel parameters for eth0:
Pre-set maximums:
RX:		0
TX:		0
Other:		0
Combined:	4
Current hardware settings:
RX:		0
TX:		0
Other:		0
Combined:	4

root@stretch-test:~# ls /sys/class/net/eth0/queues/
rx-0  rx-1  rx-2  rx-3	tx-0  tx-1  tx-2  tx-3
```

However, for some strange reason the VM's internet connection is messed-up. Digging in Ganeti's codebase we noticed [this](https://github.com/ganeti/ganeti/blob/47eca2c618107b09c8474539963fb05f0c692003/lib/hypervisor/hv_kvm/netdev.py#L171) in `OpenTap`:
```python
  for _ in range(virtio_net_queues):
    try:
      tapfd = os.open("/dev/net/tun", os.O_RDWR)
    except EnvironmentError:
      raise errors.HypervisorError("Failed to open /dev/net/tun")

    flags = IFF_TAP | IFF_NO_PI

    if vnet_hdr and _ProbeTapVnetHdr(tapfd):
      flags |= IFF_VNET_HDR

    # Check if it's ok to enable IFF_MULTI_QUEUE
    if virtio_net_queues > 1 and _ProbeTapMqVirtioNet(tapfd):
      flags |= IFF_MULTI_QUEUE
    else:
      flags |= IFF_ONE_QUEUE

    # The struct ifreq ioctl request (see netdevice(7))
    ifr = struct.pack("16sh", name, flags)

    try:
      res = fcntl.ioctl(tapfd, TUNSETIFF, ifr)
    except EnvironmentError, err:
      raise errors.HypervisorError("Failed to allocate a new TAP device: %s" %
                                   err)

    tapfds.append(tapfd)

  # Get the interface name from the ioctl
  ifname = struct.unpack("16sh", res)[0].strip("\x00")
```
So, Ganeti creates and prepares the taps correctly and then returns the **last** tap as the interface name. However, QEMU by default uses the first `fd` to initialize the NIC device and the rest only when they are enabled. Thus, at first boot with multiqueues, Ganeti runs all the [`ConfigureNIC`](https://github.com/ganeti/ganeti/blob/master/lib/hypervisor/hv_kvm/__init__.py#L874) hooks for the last tap and QEMU enables the first one. A proof-of-concept hotfix solved the connectivity issues for the first queue:
```diff
diff --git a/lib/hypervisor/hv_kvm/netdev.py b/lib/hypervisor/hv_kvm/netdev.py
index 852d48f..1810684 100644
--- a/lib/hypervisor/hv_kvm/netdev.py
+++ b/lib/hypervisor/hv_kvm/netdev.py
@@ -149,6 +149,7 @@ def OpenTap(name="", features=None):
   vhost = features.get("vhost", False)
   vnet_hdr = features.get("vnet_hdr", True)
   _, virtio_net_queues = features.get("mq", (False, 1))
+  res = []

   for _ in range(virtio_net_queues):
     try:
@@ -171,7 +172,7 @@ def OpenTap(name="", features=None):
     ifr = struct.pack("16sh", name, flags)

     try:
-      res = fcntl.ioctl(tapfd, TUNSETIFF, ifr)
+      res.append(fcntl.ioctl(tapfd, TUNSETIFF, ifr))
     except EnvironmentError, err:
       raise errors.HypervisorError("Failed to allocate a new TAP device: %s" %
                                    err)
@@ -190,6 +191,6 @@ def OpenTap(name="", features=None):
     tapfds.append(tapfd)

   # Get the interface name from the ioctl
-  ifname = struct.unpack("16sh", res)[0].strip("\x00")
+  ifname = struct.unpack("16sh", res[0])[0].strip("\x00")

   return (ifname, tapfds, vhostfds)
```
However, the other queues are not working because the network hooks were not run for the rest of the taps. The state of one queue working and the others not working resulted in "strange" network behavior e.g. pings working partialy, etc. A little of `tcpdump`ing on the host and the guest devices helped a lot! Adding the rest of the taps in the host's bridge *manually* solved the issue:
```
root@ena:~# gnt-network list
Network Subnet           Gateway       MacPrefix GroupList                      Tags
vlan992 XXX               XXX  -         default (bridged, vlan992, )   bridged,dns

root@ena:~# brctl show
bridge name	bridge id		STP enabled	interfaces
vlan992		8000.00262d0062f8	no		bond0.992
							tap0
							tap1
							tap2
							tap3
```
and resolved in normal traffic in the VM:
```
root@stretch-test:~# cat /proc/interrupts 
           CPU0       CPU1       CPU2       CPU3       
< snip >
 24:          0          0          0          0   PCI-MSI 212992-edge      virtio2-config
 25:     178854          0          0          0   PCI-MSI 212993-edge      virtio2-input.0
 26:          2          0          0          0   PCI-MSI 212994-edge      virtio2-output.0
 27:          1     122235          0          0   PCI-MSI 212995-edge      virtio2-input.1
 28:          1          0          0          0   PCI-MSI 212996-edge      virtio2-output.1
 29:          1          0     159722          0   PCI-MSI 212997-edge      virtio2-input.2
 30:          1          0          0          0   PCI-MSI 212998-edge      virtio2-output.2
 31:          0          0          0          0   PCI-MSI 212999-edge      virtio2-input.3
 32:          0          0          0          0   PCI-MSI 213000-edge      virtio2-output.3
 33:          0          0          0          0   PCI-MSI 196608-edge      virtio1-config
 34:       9358          0          0          0   PCI-MSI 196609-edge      virtio1-req.0
< snip >
```

We tried to work on a proper fix but this seemed like an intrusive change on Ganeti's core logic and we thought it would be better to start a discussion here.

Bonus
---

Throughout the investigation, the following commands were very helpful:
- Finding the tap of the host:
```
root@ena:~# cat /var/run/ganeti/kvm-hypervisor/nic/stretch-test/0
tap3
```

- Finding the VM's pid (`414`) and open `fds` (`fds=10:12:14:16,vhostfds=11:13:15:17,`):
```
root@ena:~# ps aux | grep --color stretch-test
pool94     414  3.5  1.2 5150416 598628 ?      Sl   Sep13  43:37 qemu-system-x86_64 -enable-kvm -name stretch-test -m 512 -smp 4 -pidfile /var/run/ganeti/kvm-hypervisor/pid/stretch-test -balloon virtio -daemonize -D /var/log/ganeti/kvm/stretch-test.log -machine pc-i440fx-2.8 -monitor unix:/var/run/ganeti/kvm-hypervisor/ctrl/stretch-test.monitor,server,nowait -serial unix:/var/run/ganeti/kvm-hypervisor/ctrl/stretch-test.serial,server,nowait -usb -usbdevice tablet -vnc :5190 -uuid a2a176ee-4426-4029-9919-8c97d85d4929 -global kvm-pit.lost_tick_policy=discard -rtc base=utc,driftfix=slew -netdev type=tap,id=nic-8180bcbd-c4ea-42d0,fds=10:12:14:16,vhostfds=11:13:15:17,vhost=on -device virtio-net-pci,bus=pci.0,id=nic-8180bcbd-c4ea-42d0,addr=0xd,netdev=nic-8180bcbd-c4ea-42d0,mac=aa:00:00:e6:d5:7e,mq=on,vectors=10 -qmp unix:/var/run/ganeti/kvm-hypervisor/ctrl/stretch-test.qmp,server,nowait -qmp unix:/var/run/ganeti/kvm-hypervisor/ctrl/stretch-test.kvmd,server,nowait -boot c -device virtio-blk-pci,bus=pci.0,id=disk-d4e32ca4-59f1-416e,addr=0xc,drive=disk-d4e32ca4-59f1-416e -drive file=/var/run/ganeti/instance-disks/stretch-test:0,format=raw,if=none,cache=none,id=disk-d4e32ca4-59f1-416e -S -runas pool94
```

- Finding the taps of the VM from the fds:
```
root@ena:~# ls -la /proc/414/fd
total 0
dr-x------ 2 root   root     0 Sep 13 15:46 .
dr-xr-xr-x 9 pool94 nogroup  0 Sep 13 15:46 ..
lrwx------ 1 root   root    64 Sep 13 15:46 0 -> /dev/null
lrwx------ 1 root   root    64 Sep 13 15:46 1 -> /dev/null
lrwx------ 1 root   root    64 Sep 13 15:46 10 -> /dev/net/tun
lrwx------ 1 root   root    64 Sep 13 15:46 11 -> /dev/vhost-net
lrwx------ 1 root   root    64 Sep 13 15:46 12 -> /dev/net/tun
lrwx------ 1 root   root    64 Sep 13 15:46 13 -> /dev/vhost-net
lrwx------ 1 root   root    64 Sep 13 15:46 14 -> /dev/net/tun
lrwx------ 1 root   root    64 Sep 13 15:46 15 -> /dev/vhost-net
lrwx------ 1 root   root    64 Sep 13 15:46 16 -> /dev/net/tun
lrwx------ 1 root   root    64 Sep 13 15:46 17 -> /dev/vhost-net
< snip >

root@ena:~# cat /proc/414/fdinfo/10
pos:	120
flags:	0104002
mnt_id:	18
iff:	tap0
root@ena:~# cat /proc/414/fdinfo/12
pos:	0
flags:	0104002
mnt_id:	18
iff:	tap1
root@ena:~# cat /proc/414/fdinfo/14
pos:	0
flags:	0104002
mnt_id:	18
iff:	tap3
root@ena:~# cat /proc/414/fdinfo/16
pos:	0
flags:	0104002
mnt_id:	18
iff:	tap4
```

**Comment by atta** [comment]
Hi *,

i am not 100% sure but for me it looks like the instance sould have only one tap with N-queues

like shown in the picture http://www.linux-kvm.org/page/Multiqueue

**Comment by yiannist** [comment]
Yes, you are right! Using kvm _without_ ganeti e.g.:
```
kvm -m 2048 -vnc :0 -drive file=/srv/disk0,format=raw,cache=none,if=virtio -netdev type=tap,id=netdev0,vhost=on,queues=2 -device virtio-net-pci,mq=on,vectors=6,mac=aa:00:00:de:ad:de,netdev=netdev0
```
creates only **one** tap interface:
```
12: tap0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master vlan992 state UNKNOWN group default qlen 1000
    link/ether 12:c3:87:0c:94:08 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::10c3:87ff:fe0c:9408/64 scope link 
       valid_lft forever preferred_lft forever
```
with 2 queues (you will need to enable them from inside the guest with `ethtool -l eth0 combined 2`):
```
root@ena:~# ls /sys/class/net/tap0/queues/
rx-0  rx-1  tx-0  tx-1
```
I think this was also a bug in 47eca2c (probably because of trying to re-use some code). I will try take another look at this (unless someone beats me to it)!

**Comment by gedia** [comment]
Turns out the solution is a lot simpler: one needs to provide a name for the tap device when adding additional queues. From https://www.kernel.org/doc/Documentation/networking/tuntap.txt:

> The device allocation is the same as before, and if user wants to create multiple queues, TUNSETIFF with the same device name must be called many times with IFF_MULTI_QUEUE flag.

So we just need to set "name" after the first iteration of the loop found in the function, if it's not already set:
```
--- a/lib/hypervisor/hv_kvm/netdev.py
+++ b/lib/hypervisor/hv_kvm/netdev.py
@@ -189,6 +189,9 @@
 
     tapfds.append(tapfd)
 
+    if ( _ == 0 and name == "" ):
+        name = struct.unpack("16sh", res)[0].strip("\x00")
+
   # Get the interface name from the ioctl
   ifname = struct.unpack("16sh", res)[0].strip("\x00")
```

I'd create a pull request, but I doubt anyone is looking at them these days...

**Comment by atta** [comment]
hi i just found some time to work on it i'am about to add the patch in our test-cluster

**Comment by apoikos** [comment]
Fixed in 2.16.1 with #1326.
